flowetl/tests/integration/conftest.py

```diff
@@ -0,0 +1,168 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# -*- coding: utf-8 -*-
+"""
+Conftest for flowetl integration tests
+"""
+import os
+import shutil
+
+from time import sleep
+from subprocess import DEVNULL, Popen
+from pendulum import now, Interval
+from airflow.models import DagRun
+
+import pytest
+import docker
+
+
+@pytest.fixture(scope="session")
+def docker_client():
+    """
+    docker client object
+    """
+    return docker.from_env()
+
+
+@pytest.fixture(scope="session")
+def flowetl_tag():
+    """
+    Get flowetl tag to use
+    """
+    return os.environ.get("FLOWETL_TAG", "latest")
+
+
+# pylint: disable=redefined-outer-name
+@pytest.fixture(scope="module")
+def flowetl_container(docker_client, flowetl_tag):
+    """
+    Fixture that starts a running flowetl container and
+    yeilds the container object.
+    """
+    container = docker_client.containers.run(
+        f"flowminder/flowetl:{flowetl_tag}", detach=True
+    )
+    # breif sleep to wait for backing DB to be ready
+    sleep(2)
+    yield container
+    container.kill()
+    container.remove()
+
+
+@pytest.fixture(scope="module")
+def airflow_local_setup_mdl_scope():
+    """
+    Init the airflow sqlitedb and start the scheduler with minimal env.
+    Clean up aftwards by removing the AIRFLOW_HOME and stopping the
+    scheduler. It is neccesery to set the AIRFLOW_HOME env variable on
+    test invocation otherwise it gets created somewhere else...
+    """
+    extra_env = {
+        "AIRFLOW__CORE__DAGS_FOLDER": "./dags",
+        "AIRFLOW__CORE__LOAD_EXAMPLES": "false",
+    }
+    env = {**os.environ, **extra_env}
+    initdb = Popen(
+        ["airflow", "initdb"], shell=False, stdout=DEVNULL, stderr=DEVNULL, env=env
+    )
+    initdb.wait()
+
+    with open("scheduler.log", "w") as f:
+        scheduler = Popen(
+            ["airflow", "scheduler"], shell=False, stdout=f, stderr=f, env=env
+        )
+
+    sleep(2)
+
+    yield
+
+    scheduler.terminate()
+
+    shutil.rmtree(env["AIRFLOW_HOME"])
```

Comment:
I'm confused why it's fine to remove this folder here, but it isn't explicitly created anywhere beforehand? How will subsequent tests make use of this?

Reply:
Also, what is the purpose of `AIRFLOW_HOME` anyway / what is it used for?

Reply:
Airflow creates the directory as soon as you import anything from it - if you don't specify `AIRFLOW_HOME` then gets created in `~` perhaps that's fine? Could explicitly create the dir?

Reply:
It also needs to exist for `airflow initdb` to create the sqlite db in a non-standard location.

Reply:
Yeah, if it's easy to do then it would be much clearer to explicitly create the dir. Makes it much easier for the reader to understand what `flowetl` is doing without having to know about airflow-specific behaviour (which if you don't know it already is difficult to guess and then hunt down documentation for).

Reply:
Cool will do!

Reply:
It's also not quite clear to me whether this is running inside or outside the container? I'm always a bit hesitant to remove random directories on the host.

Also, if this is running outside the container then we should probably add a fixture which checks that `AIRFLOW_HOME` is set - which is probably a good idea anyway, because we don't want any accidentally undefined behaviour coming from tests that assume the env var exists but which might get an empty value.

(Although I realise we'll get a `KeyError` if it's not set because we're accessing it via `os.environ["AIRFLOW_HOME"]`. But it's too easy to inadvertently access it via `os.getenv("AIRFLOW_HOME")`, so probably better be safe than sorry.)

Reply:
Although I think it will already exist before you even get in to that fixture...

Reply:
Yea so add `assert False` on L96 of the conftest then run `AIRFLOW_HOME=./test_airflow_home pipenv run pytest --pdb tests/integration/test_dag_run.py` - when in pdb you will see that the dir has already been made due to imports from airflow...

Reply:
So it would be completely cosmetic and would need me to delete the dir before creating it anyway?

Reply:
Which is perhaps more confusing?

Reply:
> It's also not quite clear to me whether this is running inside or outside the container? I'm always a bit hesitant to remove random directories on the host.
> 
> Also, if this is running outside the container then we should probably add a fixture which checks that `AIRFLOW_HOME` is set - which is probably a good idea anyway, because we don't want any accidentally undefined behaviour coming from tests that assume the env var exists but which might get an empty value.
> 
> (Although I realise we'll get a `KeyError` if it's not set because we're accessing it via `os.environ["AIRFLOW_HOME"]`. But it's too easy to inadvertently access it via `os.getenv("AIRFLOW_HOME")`, so probably better be safe than sorry.)

None of the tests (other than checking that fixuid works) are happening in the container...What would a fixture that checks if it's set look like? Just something that raises an exception if not? That doesn't seem like what fixtures are for to me?

Reply:
https://github.com/MobileDynasty/pytest-env perhaps?